BDArmory/FX/ExplosionFX.cs

```diff
@@ -726,210 +792,269 @@ private void ExecutePartBlastEvent(PartBlastHitEvent eventToExecute)
-                            thickness += eventToExecute.IntermediateParts.Select(p => p.Item3).Sum(); //add armor thickness of intervening parts, if any
-                            if (BDArmorySettings.DEBUG_ARMOR) Debug.Log($"[BDArmory.ExplosionFX]: Effective Armor thickness from intermediate parts: {thickness}");
-                            float penetration = 0;
-                            var Armor = part.FindModuleImplementing<HitpointTracker>();
-                            if (Armor != null)
+                            if (shapedEffect && ((warheadType == WarheadTypes.ShapedCharge) ? (realDistance <= SCRange) : warheadType == WarheadTypes.ContinuousRod))
                             {
-                                float Ductility = Armor.Ductility;
-                                float hardness = Armor.Hardness;
-                                float Strength = Armor.Strength;
-                                float safeTemp = Armor.SafeUseTemp;
-                                float Density = Armor.Density;
-                                float vFactor = Armor.vFactor;
-                                float muParam1 = Armor.muParam1;
-                                float muParam2 = Armor.muParam2;
-                                float muParam3 = Armor.muParam3;
-                                int type = (int)Armor.ArmorTypeNum;
-
-                                //penetration = ProjectileUtils.CalculatePenetration(Caliber, Caliber, warheadType == WarheadTypes.ShapedCharge ? Power / 2 : ProjMass, ExplosionVelocity, Ductility, Density, Strength, thickness, 1);
-                                penetration = ProjectileUtils.CalculatePenetration(Caliber, warheadType == WarheadTypes.ShapedCharge ? 4000f : ExplosionVelocity, warheadType == WarheadTypes.ShapedCharge ? Power * 0.0555f : ProjMass, 1f, Strength, vFactor, muParam1, muParam2, muParam3);
-                                penetrationFactor = ProjectileUtils.CalculateArmorPenetration(part, penetration, thickness);
-
-                                if (RA != null)
+                                //float HitAngle = Vector3.Angle((eventToExecute.HitPoint + rb.velocity * TimeIndex - Position).normalized, -eventToExecute.Hit.normal);
+                                //float anglemultiplier = (float)Math.Cos(Math.PI * HitAngle / 180.0);
+                                float anglemultiplier = Vector3.Dot((eventToExecute.HitPoint + rb.velocity * TimeIndex - Position).normalized, -eventToExecute.Hit.normal);
+                                float thickness = ProjectileUtils.CalculateThickness(part, anglemultiplier);
+                                if (BDArmorySettings.DEBUG_ARMOR) Debug.Log($"[BDArmory.ExplosionFX]: Part {part.name} hit by {warheadType}; {Math.Acos(anglemultiplier) * 180f / Math.PI} deg hit, armor thickness: {thickness}");
```

Comment:
Use `Mathf.Rad2Deg` instead of `180f / Math.PI`. Also, `Mathf.Acos` instead of `Math.Acos` should avoid implicitly casting `anglemultiplier` to `double`.
Similarly, below when setting `cosAngleOfEffect`.